src/tiled/automapper.cpp

```diff
@@ -712,8 +712,18 @@ static bool compareLayerTo(const TileLayer *setLayer,
                 bool matchListYes = false;
                 bool matchListNo  = false;
 
-                if (!setLayer->contains(x + offset.x(), y + offset.y()))
-                    return false;
+
+                if (!setLayer->contains(x + offset.x(), y + offset.y())) {
+                    foreach (const TileLayer *comparedTileLayer, listYes) {
+                        if (!comparedTileLayer->contains(x, y))
+                            return false;
```

Comment:
> However, it did not work, and I have no idea why.

I also find this hard to understand, so I started making a KV map of all boolean conditions.
However the answer came up when looking at the rule, which states:

* Examine a 3x3 region (as instructed by the regions layer)
* in the middle there must be a rock matching in the Ground layer (as there is a rock placed in the middle in the input_Ground layer)
* in the outer tiles there must be something other than (rock tile, empty tile).

The last condition seems to be a little confusing.
The rock is excluded because of the exception mentioned in the comment above the function. Essentially to deal with situations like two adjacent rocks and then not putting flowers over each other. The empty tile is excluded because we need to have 'something' there for matching.

When using your suggestion (3 times, for c1 as well as for c2 in the loops below), you can archive the corner cases in another simple different way by adding an 'inputnot_Ground' layer to the rule and put rocks in the outer tiles.
Then you remove the condition for having 'something' around.